src/probnum/linops/_linear_operator.py

```diff
@@ -446,6 +548,93 @@ def _trace_fallback(self) -> np.number:
 
         return trace
 
+    ####################################################################################
+    # Matrix Decompositions
+    ####################################################################################
+
+    def cholesky(self, lower: bool = True) -> LinearOperator:
+        r"""Computes a Cholesky decomposition of the :class:`LinearOperator`.
+
+        The Cholesky decomposition of a symmetric positive-definite matrix :math:`A \in
+        \mathbb{R}^{n \times n}` is given by :math:`A = L L^T`, where the unique
+        Cholesky factor :math:`L \in \mathbb{R}^{n \times n}` of :math:`A` is a lower
+        triangular matrix with a positive diagonal.
+
+        As a side-effect, this method will set the value of :attr:`is_positive_definite`
+        to :obj:`True`, if the computation of the Cholesky factorization succeeds.
+        Otherwise, :attr:`is_positive_definite` will be set to :obj:`False`.
+
+        The result of this computation will be cached. If :meth:`cholesky` is first
+        called with ``lower=True`` first and afterwards with ``lower=False`` or
+        vice-versa, the method simply returns the transpose of the cached value.
+
+        Parameters
+        ----------
+        lower :
+            If this is set to :obj:`False`, this method computes and returns the
+            upper triangular Cholesky factor :math:`U = L^T`, for which :math:`A = U^T
+            U`. By default (:obj:`True`), the method computes the lower triangular
+            Cholesky factor :math:`L`.
+
+        Returns
+        -------
+        cholesky_factor :
+            The lower or upper Cholesky factor of the :class:`LinearOperator`, depending
+            on the value of the parameter ``lower``. The result will have its properties
+            :attr:`is_upper_triangular`\ /:attr:`is_lower_triangular` set accordingly.
+
+        Raises
+        ------
+        numpy.linalg.LinAlgError
+            If the :class:`LinearOperator` is not symmetric, i.e. if
+            :attr:`is_symmetric` is not set to :obj:`True`.
+        numpy.linalg.LinAlgError
+            If the :class:`LinearOperator` is not positive definite.
+        """
+        if not self.is_symmetric:
```

Comment:
What happens if the `is_symmetric` and `is_positive_definite` properties are `None`, i.e. it is unknown whether the matrix is spd? According to the docstring `.cholesky()` should still run and if it succeeds, set the properties accordingly. However if `self.is_symmetric is None`, this conditional should throw an exception, if i'm not mistaken.

Reply:
According to the `Raises` section of the docstring, the method raises an exception if `is_symmetric` is not `True`, i.e. the method will fail if it is not known that the operator is indeed symmetric.
Similarly I state that the operator raises an exception if the matrix is not positive definite, while still setting `linop.is_positive_definite = False`.
This is intentional as it can be used as a test for positive definiteness.
So I don't really see a contradiction to the docstring here. Or am I missing something?

Reply:
My mistake! We might want to be lenient here and only raise an exception if the properties are `False`. Otherwise this may lead to a lot of boilerplate in particular when prototyping. Your call.

Reply:
In fact, I think it would be more in line with Python philosophy to attempt the Cholesky decomposition if it is not known whether the linop is spd and then raise an exception should Cholesky fail.

Reply:
In principle, I agree with the second point. In fact this is what happens if `is_positive_definite` is `None`: We attempt the Cholesky and if it fails, we set it to `False` and raise an exception. However, after giving this some thought, this issue is much more complicated with `is_symmetric`.
Since matrices are rarely exactly symmetric, an automatic check of symmetry would be much too restrictive, unless we have some sort of "asymmetry tolerance", which is just another one of these arbitrary parameters to make things work. Personally, I came to the conclusion that it is better if the user has to call `.symmetrize` on the operator before calling `.cholesky` because it makes this whole symmetry business much more transparent.
Moreover, if `symmetrize` is too costly, the user can always choose to set `linop.symmetric = True` manually, even if the property is only fulfilled approximately.

Reply:
Note that NumPy and SciPy circumvent this issue by only operating on the lower triangle of the matrix.
```python
>>> a = np.array([
...     [1.0, 2.0],
...     [0.9, 1.0],
... ])
>>> np.linalg.cholesky(a)
array([[1.        , 0.        ],
       [0.9       , 0.43588989]])
>>> scipy.linalg.cholesky(a, lower=True)
array([[1.        , 0.        ],
       [0.9       , 0.43588989]])
```
IMHO, this is quite dangerous, since it can produce very subtle bugs.

Reply:
Sounds good to me! Thanks for the detailed explanation.

Reply:
If we find a good solution, I'm also fine with removing this in the future.
However, I would propose to make this a separate PR, since this would increase complexity by quite a bit.
We can always make interfaces less restrictive